Add a public `bundledPackages`/`excludedExternals` option to DEWP

tomalec opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->
This PR allows users of the Dependency Extraction Webpack Plugin "do a simple thing easy" (and complex things still possible) and exclude packages from externalizing in a simple way. It also upstream WooCommerceDEWP feature (https://github.com/woocommerce/woocommerce-admin/pull/7034). Or, in other words, publish the private feature of `BUNDLED_PACKAGES`.

It also includes a tiny clarification in the README (https://github.com/WordPress/gutenberg/pull/45948/commits/e8f3e855c805c08a89d00b26463198e053fb0cbd). The example was using the `entrypoint` for the assets file, whereas the plugin uses `output` config for that. (fixes https://github.com/WordPress/gutenberg/issues/49872)

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->
1. To allow doing a simple thing easy, as in the majority of DEWP's use-cases, I've seen `requestToExternal` is used to filter a list of packages.
2. To reduce the confusion of a non-Woo specific (at least conceptually) behavior of providing a simple list of packages to filter, not basing available in WP version of DEWP.
3. To reduce the confusion of `WooCommerceDependencyExtractionWebpackPlugin`'s `bundledPackages` not working for `@wordpress/` packages.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->
This PR changes the DEWP constructor to take `excludedExternals` option to use it to filter packages processed in `externalizeWpDeps` before any `requestToExternal` is called.
Turns the old `BUNDLED_PACKAGES` to the option's default value and exposes it as a static property to allow extending it by consumers.

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a Post or Page. -->
<!-- 2. Insert a Heading Block. -->
<!-- 3. etc. -->
Run automated tests
1. `npx wp-scripts test-unit-js --config test/unit/jest.config.js packages/dependency-extraction-webpack-plugin/`

Manual testing
1. Create a project that uses `@wordpress/components`,  `@wordpress/url`, `@wordpress/icons` and `@wordpress/interface`
2. Setup webpack config to use
```javascript
new DependencyExtractionWebpackPlugin( {
	requestToExternal( package ) { console.log( package ) },
	excludedExternals: [
		...DependencyExtractionWebpackPlugin.excludedExternals,
		'@wordpress/url'
	],
} ),
```
3. Confirm that `@wordpress/url`, `@wordpress/icons` and `@wordpress/interface` gets bundled, and none of them is logged.

## Screenshots or screencast <!-- if applicable -->

## Additional info:
1. I was considering keeping the `bundledPackages` name. This would keep the old name, and make it directly compatible with existing WooCommerce implementation. However:
	1. IMO, it's not the accurate name, as those packages may not get bundled at all. For example, if none of them is actually used & imported.
	2. The behavior is slightly different than in Woo, as the package will not reach `options.requestToExternal`, so this way we avoid breaking change.
	3. We can still have a direct migration path or alias for the old name.
2. I was trying to write tests to check that the default exclusions `/icons` and `/interface` are bundled, but didn't manage for test build to process it. I guess it's related to the monorepo setup. For `lodash` it works as expected.
3. Initially I added it explicitly as 
   ```javascript
	static excludedExternals = // ...
   ```
   but eslint fails to parse it. Probably due to the old ES version used.

----

cc @gziolo as you reviewed my latest DEWP related issues
cc @louwie17 as someone who implemented it in Woo, and could give a context about why it [eventually landed as `bundledPackages` not `excludedExternals`](https://github.com/woocommerce/woocommerce-admin/pull/7034/commits/23c2cd1229a9e000562e97b88267b12d4c342ccd).

tomalec committed:
Add a public `bundledPackages` option,

to exclude dependencies from externalizing.

tomalec committed:
Rename `bundledPackages` to `excludedExternals`,

to match more accurately to the actual behavior.

tomalec commented:
@gziolo Would you mind taking a look?

tyxla commented on packages/dependency-extraction-webpack-plugin/test/fixtures/bundled-packages/index.js:
```diff
@@ -0,0 +1,11 @@
+/**
+ * External dependencies
+ */
+import { isEmpty } from 'lodash';
```
Note that those imports might need an ESLint ignore now that we're fully removing Lodash in #52571 

```suggestion
// eslint-disable-next-line no-restricted-imports
import { isEmpty } from 'lodash';
```

tomalec replied:
Thanks for the heads-up. 

I changed `lodash` to `react-dom` in the test fixture (https://github.com/WordPress/gutenberg/pull/45948/commits/6c0863c33fc8ef6f3d47a17ba7cf679e90f1376c). I think it's more sustainable and future-proof to test against something we plan to support in the longer term instead of silencing warnings.

tomalec committed:
Fix the assets file names in the DEWP's README

Output and assets file paths and names are constructed from the `output` config, which may be different from just the `entrypoint`.

tomalec committed:
Merge branch 'trunk' into add/dewp-bundled

tomalec committed:
Test DEWP's `excludedExternals` using react

instead of lodash.
Avoid warnings and prepare for dropping lodash completely.

Addresses https://github.com/WordPress/gutenberg/pull/45948/files/e0d164967659fea61f769f876aa0c63010fbda62#r1262670308

tomalec commented:
@tyxla Thank you for the review. I merged the `trunk` here, addressed your comment, and add [a small improvement to the README](https://github.com/WordPress/gutenberg/pull/45948/commits/e8f3e855c805c08a89d00b26463198e053fb0cbd).

Would you mind taking another look?

tyxla commented:
@tomalec I mostly wanted to chime in regarding the Lodash usage as I noticed it during my audit for open PRs and issues that introduced lodash. Unfortunately, I'm not very familiar with the dependency extraction plugin so I might not be the best one to help review this one.

tomalec commented:
Ok, thanks :)

@gziolo @louwie17 Would you mind?
